### ISSUE TITLE
How are OKS' sigmas used in custom datasets?

### ISSUE THREAD

**Issue by ArnauCampanera** [ultralytics/ultralytics#22527]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello! 

I am using YOLOv8 and v11 pose estimation models in a custom dataset for bear pose recognition in ecology. We created a dataset and annotated a couple thousand images with 15 keypoints, totally different (in name and order) than those on COCO Keypoint dataset. Due to the smallness of our dataset we are initializing the model with the pre-trained weights provided by Ultralytics.

I have been trying to understand the meaning of the metrics given by Ultralytics when evaluating my trained model on my test set to choose between YOLOv8 or YOLOv11.

If OKS is used to calculate mAP@50 and mAP@50-95 metrics, how are sigmas defined or used? As I understand, they are usefull to the COCO Human Dataset keypoints. My question is, how are they used for a custom, non-human dataset? Are they kept constant? 

If not, then, are my metrics misleading?

Thank you very much!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ArnauCampanera, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started while an Ultralytics engineer reviews your question and will assist you soon.

For general references, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

For your question about OKS-based mAP with custom keypoints (e.g., bear pose with 15 keypoints different from COCO), please share a few details so we can best guide you:
-Exact Ultralytics version (`ultralytics` pip version) and model variants used (e.g., YOLO8n-pose, YOLO11n-pose)
-The training/evaluation commands you ran and full logs from validation
-Your dataset YAML (including `kpt_shape`, keypoint order, and any OKS-related settings if customized)
-A small sample of labeled images (screenshots) showing keypoint definitions and ordering
-Clarify whether you’re initializing from pretrained human-pose weights and whether you made any metric-related config changes

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
They don't have an effect when your dataset has any number other 17 as keypoints 

https://github.com/ultralytics/ultralytics/blob/c0d1d27abebc6777c9ba51777f1eb33e47bd9940/ultralytics/models/yolo/pose/val.py#L119

**Comment by ArnauCampanera** [comment]
Thank you very much for your response, @Y-T-G . Then, when using any number other than 17, it uses sigma = 1/N(keypoints).

Is mAP@50-95 also computed in steps of 5 from 50 to 95 like for object detection? I could only find mAP@50 and mAP@75. However, the code shows mAP@50-95 for pose (apart from Box).

**Comment by Y-T-G** [comment]
Yes, mAP@0.5-0.95 is calculated for pose. There's no separate function because the formula for calculation is the same as detection.

**Comment by PING-An32** [comment]
`OKS_SIGMA = (
    np.array([0.26, 0.25, 0.25, 0.35, 0.35, 0.79, 0.79, 0.72, 0.72, 0.62, 0.62, 1.07, 1.07, 0.87, 0.87, 0.89, 0.89])
    / 10.0
)`
The original OKS_SIGMA for COCO dataset assigned small values for delicate features like nose, eyes, ears. Big values for relatively large features like knees, ankles. 
As for YOLOv11, they use https://github.com/ultralytics/ultralytics/blob/73d658d46091988da85b265d66b29b3c334d60f7/ultralytics/utils/loss.py#L190 to calculate the scaleless error of key points. Because 'd' refers to the square of the Euclidean distance, and 'area' refers to the area of your bear anchor. OKS_SIGMA is used as a constant to scale the scaleless error. I believe you should assign small values to the delicate features of bear and vice versa, just like COCO did. And make sure the sum of your OKS_SIGMA is close to 1.
Actually if you take a closer look, the key points loss is then multiplied by a hyper-parameter, and the product is the pose loss.
https://github.com/ultralytics/ultralytics/blob/73d658d46091988da85b265d66b29b3c334d60f7/ultralytics/utils/loss.py#L558
My opinion.😄

**Comment by glenn-jocher** [comment]
Nice breakdown from @PING-An32; to add a bit of detail from the code side: for any pose dataset where `kpt_shape` is not `[17, 3]` (like your 15‑keypoint bears), both validation and training use a uniform sigma vector `1 / N_keypoints` rather than the COCO `OKS_SIGMA`, as shown in the `PoseValidator.init_metrics` and `kpt_iou` logic in the [PoseValidator reference](https://docs.ultralytics.com/reference/models/yolo/pose/val/) and in the `v8PoseLoss` / `KeypointLoss` implementation in the [loss docs](https://docs.ultralytics.com/reference/utils/loss/). This means your mAP@50 and mAP@50‑95 are internally consistent and fine for choosing between YOLOv8 and YOLO11 on your dataset—they’re just based on a uniform‑keypoint OKS rather than the human‑pose‑specific COCO weighting; if you want per‑keypoint weighting for bears, you’d currently need to customize the sigma vector in code (or via a small subclass) and keep that choice fixed across all experiments.

**Comment by Linksjie** [comment]
Hello ,I set sigma for my dataset in the project, but it didn't work. Could you please tell me the reason.Thank you very much!

**Comment by glenn-jocher** [comment]
In current Ultralytics pose, `sigma` isn’t read from your dataset YAML/config; it’s set inside the validator/loss code, and for any dataset whose `kpt_shape` is not `[17, 3]` it is explicitly forced to a uniform vector (`1 / nkpt`), so editing `OKS_SIGMA` (or “setting sigma in the project”) won’t have any effect unless you patch that logic for both validation and training (see `PoseValidator.init_metrics` in the [PoseValidator reference](https://docs.ultralytics.com/reference/models/yolo/pose/val/) and the `KeypointLoss`/`v8PoseLoss` sigma setup in the [loss reference](https://docs.ultralytics.com/reference/utils/loss/)); if you want help debugging why your patch isn’t taking effect, please open a new issue (this one is closed/stale) and include your exact code diff plus `yolo checks` output (or `ultralytics.utils.checks.collect_system_info()`), and if you’re in a notebook make sure you restarted the kernel after editing/installing.
